File: packages/python/src/lumalou/schedules.py
"""Strict codecs for the schedule blocks observed in the deployed web client.

These are wire representations, not hardware-verified restore operations.
Routine slots retain their original positions and step numbers; never group or
sort them when backing up a device. Unknown encodings raise instead of becoming
defaults. See docs/schedule-codecs.md for provenance and remaining ambiguity.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from enum import IntEnum
from typing import Literal

from ._generated import Alarm

Day = Literal[
    "sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"
]
DAYS: tuple[Day, ...] = (
    "sunday",
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday",
)
# Friday/Saturday are not contiguous with the other response IDs.
DAY_ROUTINE_RESPONSES: dict[Day, int] = dict(
    zip(DAYS, (0x2B, 0x2C, 0x2D, 0x2E, 0x2F, 0x90, 0x91))
)


def _integer(value: int, low: int, high: int, name: str) -> int:
    if (
        type(value) is not int and not isinstance(value, IntEnum)
    ) or not low <= value <= high:
        raise ValueError(f"{name} must be an integer from {low} to {high}")
    return value


def _payload(data: bytes, length: int) -> bytes:
    if not isinstance(data, bytes) or len(data) != length:
        raise ValueError(f"payload must be exactly {length} bytes")
    return data


def _tuple(value: tuple, length: int, name: str) -> None:
    if not isinstance(value, tuple) or len(value) != length:
        raise ValueError(f"{name} must be a tuple of {length} entries")


@dataclass(frozen=True)
class ClockTime:
    hour: int
    minute: int

    def __post_init__(self) -> None:
        _integer(self.hour, 0, 23, "hour")
        _integer(self.minute, 0, 59, "minute")


def _time(value: ClockTime | None) -> None:
    if value is not None and not isinstance(value, ClockTime):
        raise ValueError("time must be ClockTime or None")


def _encode_time(value: ClockTime | None) -> bytes:
    _time(value)
    if value is None:
        return b"\xff\xff"
    return bytes(
        (
            (value.hour // 10) << 4 | value.hour % 10,
            (value.minute // 10) << 4 | value.minute % 10,
        )
    )


def _decode_time(data: bytes) -> ClockTime | None:
    if data == b"\xff\xff":
        return None
    if any(byte >> 4 > 9 or byte & 0x0F > 9 for byte in data):
        raise ValueError("invalid BCD time (only FF FF means no time)")
    return ClockTime(*(10 * (byte >> 4) + (byte & 0x0F) for byte in data))


@dataclass(frozen=True)
class WeeklyTimes:
    """Seven Sunday-first times; None and midnight are distinct."""

    days: tuple[ClockTime | None, ...]

    def __post_init__(self) -> None:
        _tuple(self.days, 7, "week")
        for value in self.days:
            _time(value)


def encode_weekly_times(week: WeeklyTimes) -> bytes:
    if not isinstance(week, WeeklyTimes):
        raise ValueError("week must be WeeklyTimes")
    return b"".join(_encode_time(value) for value in week.days)


def decode_weekly_times(data: bytes) -> WeeklyTimes:
    _payload(data, 14)
    return WeeklyTimes(
        tuple(_decode_time(data[index : index + 2]) for index in range(0, 14, 2))
    )


@dataclass(frozen=True)
class WeeklyAlarms:
    """Seven Sunday-first alarm offsets and an uninterpreted sound nibble."""

    days: tuple[Alarm, ...]
    sound: int

    def __post_init__(self) -> None:
        _tuple(self.days, 7, "alarms")
        for value in self.days:
            _integer(value, 0, 10, "alarm")
        object.__setattr__(self, "days", tuple(Alarm(value) for value in self.days))
        _integer(self.sound, 0, 15, "sound")


def encode_weekly_alarms(alarms: WeeklyAlarms) -> bytes:
    if not isinstance(alarms, WeeklyAlarms):
        raise ValueError("alarms must be WeeklyAlarms")
    values = (*alarms.days, alarms.sound)
    return bytes(values[index] << 4 | values[index + 1] for index in range(0, 8, 2))


def decode_weekly_alarms(data: bytes) -> WeeklyAlarms:
    _payload(data, 4)
    values = tuple(value for byte in data for value in (byte >> 4, byte & 0x0F))
    return WeeklyAlarms(values[:7], values[7])


@dataclass(frozen=True)
class RoutineTask:
    """A task in its original wire slot (step numbering starts at one)."""

    step: int
    task: int

    def __post_init__(self) -> None:
        _integer(self.step, 1, 12, "step")
        _integer(self.task, 0, 11, "task")


@dataclass(frozen=True)
class DailyRoutine:
    """Time plus exactly twelve wire slots, including every zero/padding slot.

    A None time is FF FF, not 00 00. None slots encode 00; task zero with
    step one encodes 10, so an empty slot and task zero remain distinct.
    """

    time: ClockTime | None
    slots: tuple[RoutineTask | None, ...]

    def __post_init__(self) -> None:
        _time(self.time)
        _tuple(self.slots, 12, "routine slots")
        if any(
            value is not None and not isinstance(value, RoutineTask)
            for value in self.slots
        ):
            raise ValueError("routine slots must contain RoutineTask or None")

    @classmethod
    def from_steps(
        cls, time: ClockTime | None, steps: Iterable[Iterable[int]]
    ) -> DailyRoutine:
        """Build canonical slots for an explicit edit, preserving task order."""
        slots: list[RoutineTask | None] = []
        for number, values in enumerate(steps, 1):
            tasks = tuple(values)
            if not tasks:
                raise ValueError("a routine step must not be empty")
            for task in tasks:
                if len(slots) == 12:
                    raise ValueError("a routine holds at most twelve tasks")
                slots.append(RoutineTask(number, task))
        return cls(time, tuple(slots + [None] * (12 - len(slots))))


def encode_daily_routine(routine: DailyRoutine) -> bytes:
    if not isinstance(routine, DailyRoutine):
        raise ValueError("routine must be DailyRoutine")
    return _encode_time(routine.time) + bytes(
        0 if slot is None else slot.step << 4 | slot.task for slot in routine.slots
    )


def decode_daily_routine(data: bytes) -> DailyRoutine:
    _payload(data, 14)
    return DailyRoutine(
        _decode_time(data[:2]),
        tuple(
            None if byte == 0 else RoutineTask(byte >> 4, byte & 0x0F)
            for byte in data[2:]
        ),
    )


@dataclass(frozen=True)
class RoutineTaskStatus:
    """Runtime-only status: enum meanings and current-step sentinels unknown."""

    current_step: int
    task_states: tuple[int, ...]

    def __post_init__(self) -> None:
        _integer(self.current_step, 0, 255, "current step byte")
        _tuple(self.task_states, 12, "task states")
        for value in self.task_states:
            _integer(value, 0, 15, "task state nibble")


def decode_routine_task_status(data: bytes) -> RoutineTaskStatus:
    _payload(data, 7)
    return RoutineTaskStatus(
        data[0], tuple(value for byte in data[1:] for value in (byte >> 4, byte & 0x0F))
    )


def encode_routine_task_status(status: RoutineTaskStatus) -> bytes:
    """Lossless runtime-status serialization, NOT a SET command (0x68 is REQUEST)."""
    if not isinstance(status, RoutineTaskStatus):
        raise ValueError("status must be RoutineTaskStatus")
    return bytes((status.current_step,)) + bytes(
        status.task_states[index] << 4 | status.task_states[index + 1]
        for index in range(0, 12, 2)
    )

File: packages/js/src/schedules.ts
/** Strict, lossless schedule codecs; hardware restoration safety is unverified. */
import { DAY_ROUTINE } from "./generated.js";
import { integer, payload } from "./profile.js";

export const DAYS = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"] as const;
export type Day = typeof DAYS[number];
export const DAY_ROUTINE_RESPONSES: Record<Day, number> = {
  sunday: 0x2b, monday: 0x2c, tuesday: 0x2d, wednesday: 0x2e,
  thursday: 0x2f, friday: 0x90, saturday: 0x91,
};
export interface ClockTime { readonly hour: number; readonly minute: number }
export interface WeeklyTimes { readonly days: readonly (ClockTime | null)[] }
export interface WeeklyAlarms { readonly days: readonly number[]; readonly sound: number }
export interface RoutineTask { readonly step: number; readonly task: number }
export interface DailyRoutine { readonly time: ClockTime | null; readonly slots: readonly (RoutineTask | null)[] }
export interface RoutineTaskStatus { readonly currentStep: number; readonly taskStates: readonly number[] }

function entries(value: readonly unknown[], length: number): void {
  if (!Array.isArray(value) || value.length !== length) throw new Error(`expected ${length} entries`);
}

function encodeTime(value: ClockTime | null): number[] {
  if (value === null) return [255, 255];
  const hour = integer(value.hour, 0, 23, "hour"), minute = integer(value.minute, 0, 59, "minute");
  return [Math.floor(hour / 10) << 4 | hour % 10, Math.floor(minute / 10) << 4 | minute % 10];
}

function decodeTime(data: Uint8Array): ClockTime | null {
  if (data[0] === 255 && data[1] === 255) return null;
  for (const byte of data) if ((byte >> 4) > 9 || (byte & 15) > 9) throw new Error("invalid BCD time");
  const value = { hour: (data[0] >> 4) * 10 + (data[0] & 15), minute: (data[1] >> 4) * 10 + (data[1] & 15) };
  encodeTime(value);
  return value;
}

export function encodeWeeklyTimes(value: WeeklyTimes): Uint8Array {
  entries(value.days, 7);
  return Uint8Array.from(Array.from(value.days).flatMap(encodeTime));
}

export function decodeWeeklyTimes(data: Uint8Array): WeeklyTimes {
  payload(data, 14);
  return { days: Array.from({ length: 7 }, (_, index) => decodeTime(data.slice(index * 2, index * 2 + 2))) };
}

export function encodeWeeklyAlarms(value: WeeklyAlarms): Uint8Array {
  entries(value.days, 7);
  const values = [...Array.from(value.days, v => integer(v, 0, 10, "alarm")), integer(value.sound, 0, 15, "sound nibble")];
  return Uint8Array.from({ length: 4 }, (_, i) => values[i * 2] << 4 | values[i * 2 + 1]);
}

export function decodeWeeklyAlarms(data: Uint8Array): WeeklyAlarms {
  payload(data, 4);
  const values = Array.from(data).flatMap(b => [b >> 4, b & 15]);
  const value = { days: values.slice(0, 7), sound: values[7] };
  encodeWeeklyAlarms(value);
  return value;
}

export function routineFromSteps(time: ClockTime | null, steps: readonly (readonly number[])[]): DailyRoutine {
  const slots: (RoutineTask | null)[] = [];
  for (const [index, tasks] of steps.entries()) {
    if (tasks.length === 0) throw new Error("routine step must not be empty");
    for (const task of tasks) {
      if (slots.length === 12) throw new Error("routine holds at most twelve tasks");
      slots.push({ step: index + 1, task });
    }
  }
  while (slots.length < 12) slots.push(null);
  const value = { time, slots };
  encodeDailyRoutine(value);
  return value;
}

export function encodeDailyRoutine(value: DailyRoutine): Uint8Array {
  entries(value.slots, 12);
  return Uint8Array.from([...encodeTime(value.time), ...Array.from(value.slots, slot =>
    slot === null ? 0 : integer(slot.step, 1, 12, "step") << 4 | integer(slot.task, 0, 11, "task"))]);
}

export function decodeDailyRoutine(data: Uint8Array): DailyRoutine {
  payload(data, 14);
  const value = { time: decodeTime(data.slice(0, 2)), slots: Array.from(data.slice(2), byte =>
    byte === 0 ? null : { step: byte >> 4, task: byte & 15 }) };
  encodeDailyRoutine(value);
  return value;
}

/** Runtime-only. Task-state enum meanings and current-step sentinels are unknown. */
export function decodeRoutineTaskStatus(data: Uint8Array): RoutineTaskStatus {
  payload(data, 7);
  return { currentStep: data[0], taskStates: Array.from(data.slice(1)).flatMap(b => [b >> 4, b & 15]) };
}

export function encodeRoutineTaskStatus(value: RoutineTaskStatus): Uint8Array {
  entries(value.taskStates, 12);
  const values = Array.from(value.taskStates, v => integer(v, 0, 15, "task state nibble"));
  return Uint8Array.of(integer(value.currentStep, 0, 255, "current step byte"), ...Array.from({ length: 6 }, (_, i) => values[i * 2] << 4 | values[i * 2 + 1]));
}

function dayOpcode(operation: "SET" | "REQUEST", day: Day): number {
  if (!DAYS.includes(day)) throw new Error("unknown routine day");
  return DAY_ROUTINE[operation][day];
}
export const setDayRoutine = (day: Day, routine: DailyRoutine) => Uint8Array.of(dayOpcode("SET", day), ...encodeDailyRoutine(routine));
export const requestDayRoutine = (day: Day) => Uint8Array.of(dayOpcode("REQUEST", day));
export const setR2RTimes = (week: WeeklyTimes) => Uint8Array.of(0x46, ...encodeWeeklyTimes(week));
export const setSleepyTimes = (week: WeeklyTimes) => Uint8Array.of(0x48, ...encodeWeeklyTimes(week));
export const setR2RAlarms = (alarms: WeeklyAlarms) => Uint8Array.of(0x4a, ...encodeWeeklyAlarms(alarms));

export function parseScheduleResponse(opcode: number, args: Uint8Array): WeeklyTimes | WeeklyAlarms | DailyRoutine | RoutineTaskStatus {
  if (opcode === 0x22 || opcode === 0x23) return decodeWeeklyTimes(args);
  if (opcode === 0x27) return decodeWeeklyAlarms(args);
  if (Object.values(DAY_ROUTINE_RESPONSES).includes(opcode)) return decodeDailyRoutine(args);
  if (opcode === 0x94) return decodeRoutineTaskStatus(args);
  throw new Error("unsupported schedule response");
}
